### PR TITLE
[SYCLomatic] Fix header inclusion order issue in user define rule

### PR DESCRIPTION
**Pull request by yihwang-nv**
- Fix <oneapi/dpl/execution>, <oneapi/dpl/algorithm> and <dpct/dnnl_utils.hpp> includsion order issue in user define rule
- Refactor insertHeader

Signed-off-by: Wang, Yihan <yihan.wang@intel.com>

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/test/dpct/user_define_rule_header_order1.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+#include <cub/cub.cuh>
```
Comment (zhimingwang36) [inline clang/test/dpct/user_define_rule_header_order1.h:4]: In Line1,  please add check to show the behaviour of the migration

File: clang/lib/DPCT/HeaderTypes.inc [hunk 2]
```diff
@@ -0,0 +1,68 @@
+//===--------------- HeaderTypes.inc --------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef HEADER
+#define HEADER(X, Y)
```
Comment (zhimingwang36) [inline clang/lib/DPCT/HeaderTypes.inc]: Can X, Y be more meaning words?

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 3]
```diff
@@ -555,145 +555,109 @@ void DpctFileInfo::emplaceReplacements(ReplTy &ReplSet) {
     Repls->emplaceIntoReplSet(ReplSet[FilePath]);
 }
 
+static constexpr StringRef HeaderSpellings[] = {
+#define HEADER(X, Y) Y,
+#include "HeaderTypes.inc"
+};
+
+StringRef DpctFileInfo::getHeaderSpelling(HeaderType Type) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Type ==> TypeValue , or TypeEnumvalue

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 4]
```diff
@@ -555,145 +555,109 @@ void DpctFileInfo::emplaceReplacements(ReplTy &ReplSet) {
...
-                          "<oneapi/dpl/execution>");
-    case HT_DPL_Iterator:
-      return insertHeader(HeaderType::HT_DPL_Iterator, LastIncludeOffset,
-                          "<oneapi/dpl/iterator>");
-    case HT_STDLIB:
-      return insertHeader(HeaderType::HT_STDLIB, LastIncludeOffset,
-                          "<cstdlib>");
-    }
+  switch (Type) {
+#define HEADER(X, Y)                                                           \
```
Comment (zhimingwang36) [inline clang/lib/DPCT/AnalysisInfo.cpp]: X, Y ==> Give more meaningful name?

Comment (yihwang-nv) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/AnalysisInfo.cpp (hunk 3 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/AnalysisInfo.cpp]: done

File: clang/lib/DPCT/HeaderTypes.inc (hunk 2 above)
Comment (yihwang-nv) [reply clang/lib/DPCT/HeaderTypes.inc]: done

File: clang/test/dpct/user_define_rule_header_order1.h (hunk 1 above)
Comment (yihwang-nv) [reply clang/test/dpct/user_define_rule_header_order1.h:4]: done

File: clang/lib/DPCT/HeaderTypes.inc [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#else
+STD_HEADER(DL, "<dlfcn.h>")
+#endif
+
+ONEDPL_HEADER(Algorithm, "<oneapi/dpl/algorithm>")
+ONEDPL_HEADER(Execution, "<oneapi/dpl/execution>")
+ONEDPL_HEADER(Iterator, "<oneapi/dpl/iterator>")
+
+MKL_HEADER(Mkl, "<oneapi/mkl.hpp>")
+MKL_HEADER(RNG, "<oneapi/mkl/rng/device.hpp>")
```
Comment (zhiweij1) [inline clang/lib/DPCT/HeaderTypes.inc:52]: Can we include both <oneapi/mkl.hpp> and <oneapi/mkl/rng/device.hpp> for HT_MKL_RNG?

Comment (yihwang-nv) [reply clang/lib/DPCT/HeaderTypes.inc:52]: done

**Review by zhiweij1 — APPROVED** [review]

**Review by ziranzha — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**